### PR TITLE
Add GPTQ to prototype

### PR DESCRIPTION
**Pull request by jcaip**
Summary:

This PR adds GPTQ support to `torchao.prototype.gptq`.

It is exposed via a new config, GPTQConfig, which can have two steps, "observe" and "convert". 

When `quantize_(model, GPTQConfig(step="observe"))` is run, observer tensors are attached to the weight tensors, which keep track of linear / torch.bmm ops and updates the hessian matrix based on the observed inputs. 

When `quantize_(model, GPTQConfig(step="convert"))` is run, we will find any observer tensors, take the Hessian and do int4 GPTQ quantization to find the weights. The core of this function is in `gptq_quantize`. 

Currently Int4WeightOnlyConfig and Int8WeightOnlyConfig are supported. 

Also included is an example script, `gptq_example.py` that does sequential / nonsequential quantization on helllaswag for a simple example. 

I am comparing against this survey paper: https://arxiv.org/pdf/2409.11055v1 We look to be on par accuraccy wise for int8 / int4 GPTQ.  (GPTQ** is llm-compressor, GPTQ* is autogptq)

To reproduces these results please run:
```
python torchao/prototype/gptq/gptq_example.py --quantization baseline
python torchao/prototype/gptq/gptq_example.py --quantization int4-rtn
python torchao/prototype/gptq/gptq_example.py --quantization int4-gptq-sequential
python torchao/prototype/gptq/gptq_example.py --quantization int4-gptq-nonsequential
python torchao/prototype/gptq/gptq_example.py --quantization int8-rtn
python torchao/prototype/gptq/gptq_example.py --quantization int8-gptq-sequential
python torchao/prototype/gptq/gptq_example.py --quantization int8-gptq-nonsequential

```
<html><head></head><body><p>Benchmark Results:</p>

Model | Method | W/A | Storage (GB) | BBH (3-shot) | TorchAO Results
-- | -- | -- | -- | -- | --
Llama-3.1-8B-it | FP16 | 16 / 16 | 16 | 51.35 | 51.08
  | FP8 | 8 / 8 | 8 | 50.78 (↓0.57) |  
  | INT8 | 8 / 16 | 8 | N/A | 41.82
  | INT4 | 4 / 16 | 4 | N/A | 50.91
  | GPTQ* | 4 / 16 | 4 | 48.27 (↓3.08) |  
  | GPTQ** | 4 / 16 | 4 | 48.48 (↓2.87) | 48.32
  | GPTQ** | 8 / 16 | 8 | 51.29 (↑0.06) | 51.10
  | SmoothQuant | 8 / 8 | 8 | 51.32 (↓0.03) |  
  | AWQ | 4 / 16 | 4 | 48.25 (↓3.10) |  

</body></html>

Timings:

running `python torchao/prototype/gptq/gptq_example.py --quantization int4-gptq-sequential` yields a quantized model in 6.6 minutes, while running the llmcompressor example [here](https://github.com/vllm-project/llm-compressor/blob/main/examples/quantization_w4a16/llama3_example.py) takes 8 minutes, so we're approximately ~13% faster. 

Need to profile the runs to get a sense of where the speedup is happening

Test Plan:

```
pytest test/prototype/gptq/test_gptq.py 
```

### CODE REVIEW

File: test/prototype/gptq/test_gptqv2.py [hunk 1]
```diff
@@ -0,0 +1,453 @@
+# Copyright (c) Meta Platforms, Inc. and affiliates.
```
Comment (vkuzo) [inline test/prototype/gptq/test_gptqv2.py:1]: nit: put in `test/prototype/gptq` folder to match codebase convention

File: torchao/prototype/gptq/__init__.py [hunk 2]
```diff
@@ -0,0 +1,269 @@
...
+
+from torchao.core.config import AOBaseConfig
+from torchao.quantization import Int4Tensor
+from torchao.quantization.quant_api import _module_extra_repr
+from torchao.quantization.transform_module import register_quantize_module_handler
+
+from .observer import ObserverTensor
+
+
+@dataclass
```
Comment (vkuzo) [inline torchao/prototype/gptq/__init__.py]: nit: put logic in `some_file.py` instead of `__init__.py`

File: torchao/prototype/gptq/__init__.py [hunk 3]
```diff
@@ -0,0 +1,269 @@
...
+    H[dead, dead] = 1
+    W[:, dead] = 0
+
+    damp = percdamp * torch.mean(torch.diag(H))
+    diag = torch.arange(columns, device=device)
+    H[diag, diag] += damp
+    H = torch.linalg.cholesky(H)
+    H = torch.cholesky_inverse(H)
+    H = torch.linalg.cholesky(H, upper=True)
+    Hinv = H
```
Comment (vkuzo) [inline torchao/prototype/gptq/__init__.py]: nit: add some comments linking to the right page of the paper

File: torchao/prototype/gptq/__init__.py [hunk 4]
```diff
@@ -0,0 +1,269 @@
...
+
+    all_qparams = []
+
+    for W_quantize_block, block_start in zip(
```
Comment (vkuzo) [inline torchao/prototype/gptq/__init__.py]: nit: add comments as needed to explain how this code is following the algorithm in the paper

File: torchao/prototype/gptq/__init__.py [hunk 5]
```diff
@@ -0,0 +1,269 @@
...
+        shape=W.shape,
+        act_pre_scale=None,
+    )
+    return res
+
+
+def _calculate_hessian(inputs, device=None):
+    """Calculate Hessian matrix from input activations for GPTQ.
+
+    DEPRECATED: This function is kept for backward compatibility in tests only.
```
Comment (vkuzo) [inline torchao/prototype/gptq/__init__.py]: backward compatibility with what?

File: torchao/prototype/gptq/observer.py [hunk 6]
```diff
@@ -0,0 +1,96 @@
...
+# This source code is licensed under the BSD 3-Clause license found in the
+# LICENSE file in the root directory of this source tree.
+
+import torch
+import torch.nn.functional as F
+
+from torchao.utils import TorchAOBaseTensor
+
+
+class ObserverTensor(TorchAOBaseTensor):
```
Comment (vkuzo) [inline torchao/prototype/gptq/observer.py]: the logic here is specific to GPTQ, can we ensure the name specifies that

**Comment by jerryzh168** [comment]
can you show e2e accuracy results as well

**Comment by jcaip** [comment]
> I guess with this it's good that we don't need MultiTensor stuff anymore. I do remember there were some inplace op considerations before:
> 
> https://github.com/pytorch/ao/blob/f59a19bf0d97fd3672bf33497fd0b3508ec13254/torchao/quantization/GPTQ/GPTQ.py#L349
> 
> how is it addressed here?

The in-place op considerations are specific to MultiTensor, so if we avoid multi tensor we don't have to worry about them. 

We needed Multi-tensor before to support propagating quantized outputs to the subsequent observer modules. Right now we instead quantize one transformer layer at a time via some custom code in gptq_example.py. This mirrors the approach that https://github.com/IST-DASLab/MoE-Quant takes. 

llm comperssor does this by tracing the fx graph, it is possible to do something similar if we want.

File: torchao/prototype/gptq/api.py [hunk 7]
```diff
@@ -0,0 +1,330 @@
...
+
+CONFIG_TO_TORCHAO_BASE_TENSOR = {
+    Int4WeightOnlyConfig: Int4Tensor,
+    Int8WeightOnlyConfig: Int8Tensor,
+}
+
+
+@dataclass
+class GPTQConfig(AOBaseConfig):
+    """Config for GPTQ quantization
```
Comment (jcaip) [inline torchao/prototype/gptq/api.py:42]: add explanation that this is with unquantized activations out of the box

File: test/quantization/quantize_/workflows/int8/test_int8_tensor.py [hunk 8]
```diff
@@ -308,6 +308,58 @@ def test_static_activation_per_row_int8_weight(self, granularity, dtype):
         # for compile, we can't compare dynamic vs static because we may get slightly different qparams when fused
         torch.testing.assert_close(dynamic_out_eager, static_out_eager)
 
+    @common_utils.parametrize("dtype", [torch.bfloat16])
+    def test_int8_weight_only_v1_vs_v2_comparison(self, dtype):
```
Comment (vkuzo) [inline test/quantization/quantize_/workflows/int8/test_int8_tensor.py]: just test v2 and remove v1?

File: torchao/prototype/gptq/gptq_example.py [hunk 9]
```diff
@@ -0,0 +1,368 @@
...
+from torchao.prototype.gptq import GPTQConfig
+from torchao.quantization import Int4WeightOnlyConfig, Int8WeightOnlyConfig, quantize_
+from torchao.quantization.granularity import PerRow
+
+
+def sequential_quantize(
+    model,
+    calibration_data: List[torch.Tensor],
+    config: Any,
+) -> None:
```
Comment (vkuzo) [inline torchao/prototype/gptq/gptq_example.py:40]: add docblock explaining that user is expected to implement this for their model

File: torchao/prototype/gptq/gptq_example.py [hunk 10]
```diff
@@ -0,0 +1,368 @@
...
+    # run with no grad otherwise keeping all the tensors around for the backwards will cause oom
+    with torch.no_grad():
```
Comment (vkuzo) [inline torchao/prototype/gptq/gptq_example.py]: can be moved to a decorator on `sequential_quantize` to save a level of indent

File: torchao/prototype/gptq/api.py [hunk 11]
```diff
@@ -0,0 +1,330 @@
+# Copyright (c) Meta Platforms, Inc. and affiliates.
```
Comment (namgyu-youn) [inline torchao/prototype/gptq/api.py:1]: nit: rename to `api.py`, instead of `gptq.py` to following AWQ/SmoothQuant patterns?

File: torchao/prototype/gptq/gptq.py [hunk 12]
```diff
@@ -0,0 +1,330 @@
...
+# LICENSE file in the root directory of this source tree.
+
+import types
+from dataclasses import dataclass
+from functools import partial
+from typing import Union
+
+import torch
+import torch.nn as nn
+from fbgemm_gpu.experimental.gen_ai.quantize import int4_row_quantize_zp, pack_int4
```
Comment (namgyu-youn) [inline torchao/prototype/gptq/gptq.py]: Can we focus on W8A8-INT only, instead of W4A16-INT for early land? fbgemm dependency issue has been reported many times I remember.

File: torchao/prototype/gptq/api.py [hunk 13]
```diff
@@ -0,0 +1,330 @@
...
+    Int4WeightOnlyConfig,
+    Int8WeightOnlyConfig,
+    _module_extra_repr,
+)
+from torchao.quantization.transform_module import register_quantize_module_handler
+from torchao.quantization.utils import get_block_size
+
+from .observer import GPTQObserverTensor
+
+CONFIG_TO_TORCHAO_BASE_TENSOR = {
```
Comment (namgyu-youn) [inline torchao/prototype/gptq/api.py:34]: Is this needed? Can we drop this?

File: torchao/prototype/gptq/api.py [hunk 14]
```diff
@@ -0,0 +1,330 @@
...
+    for chunk, scale_chunk, zero_chunk in zip(x_chunks, scales_list, zeros_list):
+        chunk_float = chunk.to(torch.float32) + 2 ** (n_bit - 1)
+        min_val = zero_chunk - scale_chunk * (2 ** (n_bit - 1))
+        dequant = chunk_float * scale_chunk + min_val
+        dequant_chunks.append(dequant)
+
+    return torch.cat(dequant_chunks, dim=-1)
+
+
+def gptq_quantize(H: torch.Tensor, W: torch.Tensor, config: GPTQConfig):
```
Comment (namgyu-youn) [inline torchao/prototype/gptq/api.py:183]: I think we should not implement quantization module inside GPTQ algorithm. To me, the main difference between AWQ/SmoothQuant/GPTQ vs. RTN is **preprocessing something** before applying RTN. 

It can be (1) activation outlier in SmoothQuant, (2) hessian inverse in GPTQ, and (3) salient weight in AWQ. To reproduce this, most of the research and we have been used calibration-flow (`lm-eval`).

Therefore, we should follow decomposed workflow:
1. Prepare: Initialization before running the algorithm; we might need linear layer for this
2. Observe: Compute "something" based on algorithm. Note that this step should based on calibration-flow
3. Store computed value inside chunk or something
4. Convert: Apply (3) to high-precision model, not quantized model
5. quantize: RTN using `quantize_(model, base_config)`

File: test/prototype/gptq/test_gptq.py [hunk 15]
```diff
@@ -0,0 +1,534 @@
...
+)
+from torchao.prototype.gptq.observer import GPTQObserverTensor
+from torchao.quantization import Int4WeightOnlyConfig, Int8WeightOnlyConfig, quantize_
+from torchao.quantization.granularity import PerRow
+
+
+def _calculate_hessian(inputs, device=None):
+    """Calculate Hessian matrix from input activations for GPTQ.
+
+    DEPRECATED: This function is kept for backward compatibility in tests only.
```
Comment (vkuzo) [inline test/prototype/gptq/test_gptq.py]: nit: update docblock

File: torchao/prototype/gptq/gptq.py [hunk 16]
```diff
@@ -0,0 +1,330 @@
...
+    """
+    This function implements the GPTQ algorithm described in this paper: https://arxiv.org/abs/2210.17323 (Algorithm 1)
+
+    GPTQ quantizes weights column-by-column while propagating quantization errors to subsequent columns.
+    This minimizes the overall error: argmin W' ||WX - W'X||_2^2
+
+    For example we can seee how GPTQ improves accuracy on a toy example:(2x2 matrix, 2-bit symmetric quantization):
+
+    Example:
+        Given: W = [[1.2, 0.8],    X = [[1.0],      Original output: W @ X = [[2.0],
```
Comment (vkuzo) [inline torchao/prototype/gptq/gptq.py]: nit: can we switch this to do `X @ W.T` to match the common case? `X @ W` is also fine, it's confusing to have `W` first here.

File: torchao/prototype/gptq/api.py [hunk 17]
```diff
@@ -0,0 +1,330 @@
...
+                            :, group_start - block_start : group_end - block_start
+                        ],
+                        base_config.granularity,
+                    )
+
+            # Quantize each column and propagate errors to subsequent columns
+            for i in range(group_start - block_start, group_end - block_start):
+                w = W_quantize_block[:, i].unsqueeze(1)
+                if isinstance(base_config, Int4WeightOnlyConfig):
+                    q = _int4_row_quantize_zp_precomputed_qparams(
```
Comment (vkuzo) [inline torchao/prototype/gptq/api.py:303]: can probably improve how int4/int8/etc logic works in these two for loops, but better in a future PR

File: torchao/prototype/gptq/gptq.py [hunk 18]
```diff
@@ -0,0 +1,330 @@
...
+                W_quantize_block[:, i:] -= err1.matmul(
+                    Hinv_quantize_block[i, i:].unsqueeze(0)
+                )
+                Err1[:, i] = err1.flatten()
+
+        # Lazy Batch-Updates: We process B columns at a time with local updates above.
+        # Once a block is fully processed, perform global updates to H^-1 and W using batched versions of the error propagation equations.
+        W[:, block_end:] -= Err1.matmul(Hinv[block_start:block_end, block_end:])
+
+    if "cuda" in device.type:
```
Comment (vkuzo) [inline torchao/prototype/gptq/gptq.py]: does this support cpu? maybe just drop cpu for now? my guess is things will be too slow to be useful on cpu

File: torchao/quantization/quantize_/workflows/int8/int8_tensor.py [hunk 19]
```diff
@@ -137,6 +137,8 @@ def from_hp(
                 scale_dtype=hp_tensor.dtype,
                 zero_point_dtype=torch.int8,
                 keepdim=True,
+                # Use float32 eps for all dtypes to avoid accuracy loss with bfloat16 (bfloat16 eps is ~65x larger).
```
Comment (vkuzo) [inline torchao/quantization/quantize_/workflows/int8/int8_tensor.py:144]: separate PR

File: torchao/prototype/gptq/api.py (hunk 14 above)
Comment (vkuzo) [reply torchao/prototype/gptq/api.py:183]: @namgyu-youn , the GPTQ algorithm doesn't cleanly translate to your steps 1-6 above without either program capture, custom model forward or other tricks.  The current PR takes a "custom model forward" approach.  I would recommend improving the usability in future PRs.

Comment (namgyu-youn) [reply torchao/prototype/gptq/api.py:183]: > @namgyu-youn , the GPTQ algorithm doesn't cleanly translate to your steps 1-6 above without either program capture, custom model forward or other tricks. The current PR takes a "custom model forward" approach. I would recommend improving the usability in future PRs.

Oh sorry I had to review OBQ again. GPTQ can't be done with 1-6 steps like AWQ/SmoothQuant. Mark as out of topic

File: torchao/prototype/gptq/gptq.py (hunk 12 above)
Comment (jcaip) [reply torchao/prototype/gptq/gptq.py]: I think this just needs to be gated properly like the other fbgemm imports

File: torchao/prototype/gptq/api.py (hunk 13 above)
Comment (jcaip) [reply torchao/prototype/gptq/api.py:34]: We need something to map from the configs to the tensor, which we don't have currently afaik.

File: torchao/quantization/quantize_/workflows/int8/int8_tensor.py (hunk 19 above)
Comment (jcaip) [reply torchao/quantization/quantize_/workflows/int8/int8_tensor.py:144]: landed in #3664

File: torchao/prototype/gptq/gptq.py (hunk 18 above)
Comment (jcaip) [reply torchao/prototype/gptq/gptq.py]: It should work but i agree this will be quite slow on CPU.
